### PR TITLE
Espnet3/installation

### PR DESCRIPTION
**Pull request by Masao-Someki**
## What?

Installation/CI fix for ESPnet-3 development. Work done by @Fhrozen 

## Why?

1. espnet-3 requires additional dependencies.
2. Since espnet installation contains so many dependencies, and some of them are not required for some tasks. We wanted to clean the dependencies based on use cases.

## See also

#6133 

## Major changes

`.devcontainer/`

- Would like to ask Nelson to explain this part, but basically creating devcontainer for ESPnet-3, and refactoring the code base.

`.github/`

- Addition of utilities to simplify environment setup for CI.

`.pre-commit-config.yaml`

- Commented out to temporarily disable functionality for automatic format execution.
- Will uncomment before merge.

`ci/`

- espnet-2: Mainly format and task additions. Want to run CI in parallel by adding tasks as arguments.
- For espnet-1 related items, changed paths to accommodate folder structure changes in test_utils.

- Removed espnet-ez related CI. (ESPnet-3 CI will be added in the coming PR)

`doc`, `vuepress`

- Homepage changes and documentation addition for devcontainer.

`egs*/espnet*`

- I'm sorry, I will delete items that is still not be reverted for formatting.
- For some changes are not related to formats, so currently asking to Nelson.

`test/` 

- Support `torch>2.6`.

`test_utils/`

- Moved espnet1 resources to `espnet1_scripts`.

`tools/` 
- torch2.7 support
- Changes to makefile to use pyproject.toml instead of setup.py.

### CODE REVIEW

**Comment by sw005320** [comment]
- I created the ESPnet3 label and added it to this PR and your roadmap. 
  - Similar to ESPnet2, we may automatically put this label in the PR. Can you do it by checking the ESPnet2 label?
- There are too many (281!) changes (mostly the format change). Please check them and remove them if they are unnecessary.

**Comment by sw005320** [comment]
- Please also change the installation document accordingly. The installation document must be located at https://espnet.github.io/espnet/installation.html
- BTW, https://espnet.github.io/espnet/ does not have a pointer to the installation step, which is the most important documentation.

**Comment by Masao-Someki** [comment]
Thank you for your comment!

>BTW, https://espnet.github.io/espnet/ does not have a pointer to the installation step, which is the most important documentation.

Thank you, I didn't realized it...!! I just reached out to @juice500ml for fixing it.

**Comment by Masao-Someki** [comment]
Hi @Fhrozen, would it be okay to ask for an edit to `installation.md` ?
I think we should at least add:
- a link to `devcontainer.md`
- pip command with tasks, such as `pip install espnet[task-asr]`

**Comment by juice500ml** [comment]
Added requested fixes for docs in https://github.com/espnet/espnet/pull/6135/commits/43b2f1bdde25a2b455edbc117d8c2539ad38cadb

"Full ESPnet Installation" link added
image

New navbar tab "Tutorials" added
image

**Comment by Masao-Someki** [comment]
I've checked all diffs and following files are just formatting the script.
- `egs/`
- `egs2/`
- `test/`: except for `test_asr_inference.py` , `test_hugging_face_transformers_decoder.py`, and `test_train_dtype.py`
- `utils/`

I will remove those diffs.

I also found `espnet-3` files are included, so I will remove it. I will include those on the following PR that contains espnet-3/tests.

**Comment by Masao-Someki** [comment]
@Fhrozen Could you check the modification on the following files are related to installation/CI?

- espnet/nets/pytorch_backend/transformer/embedding.py
- espnet2/asr/encoder/whisper_encoder.py
- espnet2/train/collate_fn.py (I think this is related to CI, just for confirmation)

**Comment by Masao-Someki** [comment]
Thank you @juice500ml!
I feel it might be better to replace this block in the former one?
https://github.com/espnet/espnet/blob/43b2f1bdde25a2b455edbc117d8c2539ad38cadb/doc/index.md?plain=1#L18-L21

**Comment by Fhrozen** [comment]
> @Fhrozen Could you check the modification on the following files are related to installation/CI?
> 
> * espnet/nets/pytorch_backend/transformer/embedding.py
> * espnet2/asr/encoder/whisper_encoder.py
> * espnet2/train/collate_fn.py (I think this is related to CI, just for confirmation)

Sure np

**Comment by juice500ml** [comment]
Discussed w Masao offline, applied his suggestions in d51d90f (change url from tutorial to installation)

**Comment by sw005320** [comment]
> Added requested fixes for docs in [43b2f1b](https://github.com/espnet/espnet/commit/43b2f1bdde25a2b455edbc117d8c2539ad38cadb)
> 
> "Full ESPnet Installation" link added image
> 
> New navbar tab "Tutorials" added image

Can you add the installation guide to the "Get started with ESPnet!" section or a similar place (you may create a new field)?
The installation part should be located in a visible place. I think an additional item in the "Get started with ESPnet!" section

**Comment by juice500ml** [comment]
@sw005320 
The third tab (High performance...) leads to the installation page, per https://github.com/espnet/espnet/commit/d51d90fb04d50a78f45c09790118b2095100f17b . Or maybe it should be more verbose?
image

**Comment by Masao-Someki** [comment]
I forgot that formatting is automatically done by precommit.ci. I just revert formatting and pushed without triggering any hooks.

@ftshijt Could you review this PR if it is okay to merge?
I will also continue working on adding espnet-3 contents based on this PR!!

**Comment by sw005320** [comment]
> @sw005320 The third tab (High performance...) leads to the installation page, per [d51d90f]

This is very indirect for me.
Can you clarify more about this description, then?

Maybe
"Full experiment replication with install guide" ???

**Comment by sw005320** [comment]
> I forgot that formatting is automatically done by precommit.ci. I just revert formatting and pushed without triggering any hooks.
> 
> @ftshijt Could you review this PR if it is okay to merge? I will also continue working on adding espnet-3 contents based on this PR!!

You still have more than 200 files.
I believe something is wrong.

**Comment by Masao-Someki** [comment]
Oh, precommit run again... It should be 85 files..

**Comment by Masao-Someki** [comment]
I just reverted formatting and turn off precommit. The files that requires review is now 85.

**Comment by Masao-Someki** [comment]
Oh, it seems precommit document is wrong? I will check more in detail on how to turn it off

**Comment by sw005320** [comment]
> Oh, it seems precommit document is wrong? I will check more in detail on how to turn it off

Formatting issues happen in all Python files under `egs` and `egs2`

**Comment by Masao-Someki** [comment]
@ftshijt @sw005320 @Fhrozen
Sorry for the formatting issues earlier, and thank you for your help, @Fhrozen.
There are now 85 files that require review. Could you please take a look at the PR?

File: ci/test_configuration_espnet2.sh [hunk 1]
```diff
@@ -2,16 +2,48 @@
...
+execute_config(){
+    log "Testing config $2"
+    if ! ${python} -m $1 --config $2 $3 &> ${logfile}; then
+        log "ERROR: Config file $2"
+        cat ${logfile}
+        exit 1
+    else
+        log "Config file $2 OK"
+    fi
+    sudo rm -rf /root/.cache/huggingface*
```
Comment (Fhrozen) [inline ci/test_configuration_espnet2.sh]: I added `rm -rf hf_cache hub` after this line to avoid Error and `data space` issues.

File: test/espnet2/bin/test_cls_inference.py [hunk 2]
```diff
@@ -12,6 +12,8 @@
 from espnet2.bin.cls_inference import Classification, get_parser, inference, main
 from espnet2.tasks.cls import CLSTask
 
+# import yaml
```
Comment (Fhrozen) [inline test/espnet2/bin/test_cls_inference.py:14]: probably, we can remove it

File: .github/workflows/ci_on_ubuntu.yml [hunk 3]
```diff
@@ -252,40 +264,87 @@ jobs:
...
+          sudo rm -rf /root/.cache/*
+      - name: test configuration
+        run: ./ci/test_configuration_espnet2.sh ${{ matrix.task }}
       - uses: codecov/codecov-action@v2
         with:
-          flags: test_integration_espnetez
-
+          flags: test_configuration_espnet2_${{ matrix.task }}
 
   test_import:
```
Comment (Fhrozen) [inline .github/workflows/ci_on_ubuntu.yml:349]: This test is making CI fail. Can you comment it? I will need to modify it later. 
Especially, with the new changes for ESPnet2 installation, it requires moving some files between directories during the installation process.  I will be activating it later

**Comment by Fhrozen** [comment]
> @ftshijt @sw005320 @Fhrozen Sorry for the formatting issues earlier, and thank you for your help, @Fhrozen. There are now 85 files that require review. Could you please take a look at the PR?

There are some files that require formatting and deleting some unused packages.

File: .github/workflows/ci_on_ubuntu.yml (hunk 3 above)
Comment (Masao-Someki) [reply .github/workflows/ci_on_ubuntu.yml:349]: I've commented line 345-347 for now.

**Comment by Masao-Someki** [comment]
@sw005320 @ftshijt 
Here's the summary on this PR.

@Fhrozen Could you please add a description of the changes in `.devcontainer` if the reviewer asks about it? Also, if any of my descriptions below are incorrect, please let me know!

## Major changes

`.devcontainer/`

- Would like to ask Nelson to explain this part, but basically creating devcontainer for ESPnet-3, and refactoring the code base.

`.github/`

- Addition of utilities to simplify environment setup for CI.


`.pre-commit-config.yaml`

- Commented out to temporarily disable functionality for automatic format execution.
- Will uncomment before merge.


`ci/`

- espnet-2: Mainly format and task additions. Want to run CI in parallel by adding tasks as arguments.
- For espnet-1 related items, changed paths to accommodate folder structure changes in test_utils.

- Removed espnet-ez related CI. (ESPnet-3 CI will be added in the coming PR)


`doc`, `vuepress`

- Homepage changes and documentation addition for devcontainer.


`egs*/espnet*`

- I'm sorry, I will delete items that is still not be reverted for formatting.
- For some changes are not related to formats, so currently asking to Nelson.

`test/` 

- Support `torch>2.6`.


`test_utils/`

- Moved espnet1 resources to `espnet1_scripts`.


`tools/` 
- torch2.7 support
- Changes to makefile to use pyproject.toml instead of setup.py.

---

Thank you very much for all your work on this. I would greatly appreciate your review!

**Comment by sw005320** [comment]
> @sw005320 @ftshijt Here's the summary on this PR.
> 
> @Fhrozen Could you please add a description of the changes in `.devcontainer` if the reviewer asks about it? Also, if any of my descriptions below are incorrect, please let me know!
> 
> ## Major changes
> `.devcontainer/`
> 
> * Would like to ask Nelson to explain this part, but basically creating devcontainer for ESPnet-3, and refactoring the code base.
> 
> `.github/`
> 
> * Addition of utilities to simplify environment setup for CI.
> 
> `.pre-commit-config.yaml`
> 
> * Commented out to temporarily disable functionality for automatic format execution.
> * Will uncomment before merge.
> 
> `ci/`
> 
> * espnet-2: Mainly format and task additions. Want to run CI in parallel by adding tasks as arguments.
> * For espnet-1 related items, changed paths to accommodate folder structure changes in test_utils.
> * Removed espnet-ez related CI. (ESPnet-3 CI will be added in the coming PR)
> 
> `doc`, `vuepress`
> 
> * Homepage changes and documentation addition for devcontainer.
> 
> `egs*/espnet*`
> 
> * I'm sorry, I will delete items that is still not be reverted for formatting.
> * For some changes are not related to formats, so currently asking to Nelson.
> 
> `test/`
> 
> * Support `torch>2.6`.
> 
> `test_utils/`
> 
> * Moved espnet1 resources to `espnet1_scripts`.
> 
> `tools/`
> 
> * torch2.7 support
> * Changes to makefile to use pyproject.toml instead of setup.py.
> 
> Thank you very much for all your work on this. I would greatly appreciate your review!

You can move this to the top message of this PR.

**Comment by Masao-Someki** [comment]
Thank you, I added the major changes section at the initial comment!

**Comment by juice500ml** [comment]
> @sw005320 The third tab (High performance...) leads to the installation page, per [d51d90f]
> This is very indirect for me. Can you clarify more about this description, then?
> Maybe "Full experiment replication with install guide" ???


@sw005320 Agreed, "Complete installation to fully reproduce models" in 7cd82d2

**Comment by Masao-Someki** [comment]
@Fhrozen Could you check the CI issue..? If my understanding is correct, those errors basically comes from the formatting?

For the issue in `generate_and_deploy_doc`, the reason is the lack of dependencies in `pyproject.toml` for docs. I asked @juice500ml to fix it.

**Comment by ftshijt** [comment]
One proposal: maybe we can remove the espnet1 support from espnet3 by removing the related CI tests? It has been 6 years since the espnet2 cames out, we might need support window to maximize our supporting efforts

**Comment by juice500ml** [comment]
There were some old configs + missing deps in new `pyproject.toml`, so that `generate_and_deploy_doc` CI were broken. Hopefully 6491fc0 and be8eb7a fixes this? (BTW, the new `pyproject.toml` looks great! 👍)

**Comment by sw005320** [comment]
Do you know what's happening?
It looks like this would be another formatting issue.
https://github.com/espnet/espnet/actions/runs/15568526717/job/43838458450?pr=6135

**Comment by Masao-Someki** [comment]
Based on the discussion with @Fhrozen, I just deactivate the two tests that stops CI. Those are espnet-2 tests, and likely to run after formatting the repo.

**Comment by sw005320** [comment]
OK, but honestly, I cannot understand it.
Why does this PR break the CI? This PR is not related to them.

Can you use git commands to make some of the files the same as the main repo?

**Comment by sw005320** [comment]
E.g., 
```
git checkout main -- path/to/dir1 path/to/dir2
```

**Comment by juice500ml** [comment]
Introducing Tutorials had problems while generating sidebars in documents. 9fb310f and 3a127bc fixes it. Hopefully the document is generated correctly now.

**Comment by Masao-Someki** [comment]
I checked the "Files changed" tab, and this PR doesn't modify the files causing the CI failure. I also ran the git command locally and didn’t find any differences.

File: espnet2/asr/encoder/whisper_encoder.py
Comment (sw005320) [inline espnet2/asr/encoder/whisper_encoder.py:1]: Why does this change happen?
This must not be related to this PR.

**Comment by sw005320** [comment]
Is it possible to create a new PR?
This PR has too many irrelevant files, and it is very difficult to proceed with it.

File: espnet2/train/collate_fn.py
Comment (sw005320) [inline espnet2/train/collate_fn.py:1]: Why does this change happen?
This must not be related to this PR.

**Comment by Masao-Someki** [comment]
It's possible to create a new PR, but we need to clearly specify what should be included..

For those particular files, I'm currently asking @Fhrozen to confirm whether the changes are necessary ([link to comment](https://github.com/espnet/espnet/pull/6135#issuecomment-2957095417)).

Aside from those files, I believe the other changes should be included..

**Comment by sw005320** [comment]
I found that 

> `test_utils/`
>
> Moved espnet1 resources to `espnet1_scripts`

has a major issue.
Some scripts are not related to espnet1.
We should check which one we can move to `espnet1_scripts` one by one.

**Comment by sw005320** [comment]
@Fhrozen, is it possible to separate this PR and `test_utils/`? 
`test_utils/` needs some considerations and we may do it later.

**Comment by Fhrozen** [comment]
@Masao-Someki @sw005320, The CI is fixed. Let me know if there is any comment or modification is required. I will be moving to another PR as soon as this is merged.

**Comment by sw005320** [comment]
@Fhrozen, can you remove `test_utils/` related from this PR?
This PR is too large and too complex as a single PR.

**Comment by Masao-Someki** [comment]
@Fhrozen I can remove the test_utils/ related work, and could you confirm the following process looks okay to you:

1. revert the following commits:
    - [update bats for espnet1](https://github.com/espnet/espnet/commit/ac2b0eaac9b8418e435616ddc5db4d32824cef5e)
    - [rename folder](https://github.com/espnet/espnet/commit/8bfdac64fa2b37e304aed5ad1b26b72fecda3274)
2. Revert the ci fix ([updates for ci](https://github.com/espnet/espnet/commit/c911767789a99126972b179483f1853a53a5c120))

Since the second one contains formatting and weird fix for whisper encoder, I think we can revert all these three commits and can remove `test_utils/espnet1_script` related works.

**Comment by Masao-Someki** [comment]
Let me just try it, if anything happens we can check it later

**Comment by Fhrozen** [comment]
@Masao-Someki Go ahead with the revert. I will check if something breaks.

**Comment by Masao-Someki** [comment]
@Fhrozen Could you take a look at the two remaining files in test_utils/? I’m having trouble identifying which commit introduced the changes.

**Comment by Fhrozen** [comment]
@Masao-Someki, just delete the the `espnet1_scripts/` from the files `test_utils/integration_test_translate_wav.bats`
and `test_utils/integration_test_asr_align_wav.bats`

**Comment by Masao-Someki** [comment]
Workflow

1. Close this PR
2. Create PR only for flake8 (including formatting) (@Fhrozen )
3. Create PR for installation (@Masao-Someki )
4. Create PR for Pytorch>2.6 support (@Masao-Someki )
5. 
x. Create PR for homepage (@Masao-Someki )
